src/pb_stub.cc

```diff
@@ -909,29 +930,186 @@ Stub::SendLogMessage(std::unique_ptr<PbLog>& log_send_message)
   std::unique_ptr<IPCMessage> log_request_msg =
       IPCMessage::Create(shm_pool_, false /* inline_response */);
   log_request_msg->Args() = log_request_shm->ShmHandle();
+  log_request_msg->Command() = PYTHONSTUB_LogRequest;
   ScopedDefer _([send_message_payload] {
     {
-      bi::scoped_lock<bi::interprocess_mutex> guard{
-          send_message_payload->log_mu};
+      bi::scoped_lock<bi::interprocess_mutex> guard{send_message_payload->mu};
       send_message_payload->waiting_on_stub = false;
-      send_message_payload->log_cv.notify_all();
+      send_message_payload->cv.notify_all();
     }
   });
 
   {
     // Send a message to be caught by the log monitor thread in python_be.cc
-    bi::scoped_lock<bi::interprocess_mutex> guard{send_message_payload->log_mu};
-    SendIPCLogMessage(log_request_msg);
+    bi::scoped_lock<bi::interprocess_mutex> guard{send_message_payload->mu};
+    SendIPCUtilsMessage(log_request_msg);
     while (!send_message_payload->waiting_on_stub) {
-      send_message_payload->log_cv.wait(guard);
+      send_message_payload->cv.wait(guard);
     }
   }
 }
 
+void
+Stub::SendCleanupId(void* id)
+{
+  std::unique_ptr<IPCMessage> ipc_message =
+      IPCMessage::Create(shm_pool_, true /* inline_response */);
+  ipc_message->Command() = PYTHONSTUB_CleanupRequest;
+  AllocatedSharedMemory<char> cleanup_request_message =
+      shm_pool_->Construct<char>(
+          sizeof(CleanupMessage) +
+          sizeof(bi::managed_external_buffer::handle_t));
+  CleanupMessage* cleanup_message_ptr =
+      reinterpret_cast<CleanupMessage*>(cleanup_request_message.data_.get());
+  cleanup_message_ptr->id = id;
+  ipc_message->Args() = cleanup_request_message.handle_;
+
+  ScopedDefer data_load_complete([&ipc_message] {
+    bi::scoped_lock<bi::interprocess_mutex> lock{
+        *(ipc_message->ResponseMutex())};
+    ipc_message->ResponseCondition()->notify_all();
+  });
+
+  {
+    bi::scoped_lock<bi::interprocess_mutex> lock{
+        *(ipc_message->ResponseMutex())};
+    SendIPCUtilsMessage(ipc_message);
+    ipc_message->ResponseCondition()->wait(lock);
+  }
+}
+
+void
+Stub::EnqueueCleanupId(void* id)
+{
+  {
+    std::lock_guard<std::mutex> guard{utils_message_mutex_};
+    bls_response_cleanup_buffer_.push(id);
+  }
+  utils_message_cv_.notify_one();
+}
+
 bool
-Stub::LogServiceActive()
+Stub::UtilsServiceActive()
 {
-  return log_thread_;
+  return utils_thread_;
+}
+
+void
+Stub::LaunchBLSResponseQueueMonitor()
+{
+  bls_response_thread_ = true;
+  bls_response_monitor_ = std::thread(&Stub::BLSResponseQueueMonitor, this);
+}
+
+void
+Stub::TerminateBLSResponseQueueMonitor()
+{
+  if (bls_response_thread_) {
+    bls_response_thread_ = false;
+    bls_response_queue_->Push(DUMMY_MESSAGE);
+    bls_response_monitor_.join();
+  }
+}
+
+void
+Stub::BLSResponseQueueMonitor()
```

Comment:
This can be a part of a future refactor but I was wondering whether we can create a class that encapsulates the listener threads in the parent and stub process side along with a message queue. Then, we can extend from this class to have specific handling for Logging/BLS resposnes.

Reply:
Great idea! Right now we don't actually need this `BLSResponseQueueMonitor` as the comment below. I think we can refactor this part maybe in the python custom metrics feature. Does it make sense?

Reply:
Sounds good.